File: apps/api_py/src/linkbook/agents/reconciler.py
"""§5.3 — Reconciler. Proposes payment.apply only at >= 0.85 confidence
AND when there is exactly one candidate. Otherwise leaves the event
manual.
"""

from __future__ import annotations

from typing import Any

from sqlalchemy.orm import Session

from ..config import AppConfig
from ..db.models import Action, AuditEvent
from ..idempotency import compute_idempotency_key, find_duplicate_in_window
from ..lib.log import AppLogger
from ..types import ACTION_CATALOG, RECONCILER_CONFIDENCE_THRESHOLD
from .runtime import ProposeRequest, get_agentspan_client, propose_with_fallback

VERSION = "v1.0"


async def run_reconciler(
    event: dict[str, Any], cfg: AppConfig, db: Session, log: AppLogger
) -> None:
    if event["type"] != "payment.received_unapplied":
        return

    p = event["payload"]
    candidates = p.get("candidate_invoice_ids") or []

    client = get_agentspan_client(cfg)
    result = await propose_with_fallback(
        client,
        ProposeRequest(
            agent="reconciler",
            agent_version=VERSION,
            event=event,
            context={"candidate_invoice_ids": candidates},
        ),
        log,
    )
    if not result.ok:
        return
    if result.proposal.confidence < RECONCILER_CONFIDENCE_THRESHOLD:
        log.info(
            {
                "confidence": result.proposal.confidence,
                "threshold": RECONCILER_CONFIDENCE_THRESHOLD,
            },
            "reconciler below threshold — leaving as manual event",
        )
        return
    # Multi-candidate auto-apply is the failure mode that breaks
    # accountant trust. Only auto-propose when exactly one candidate.
    if len(candidates) != 1:
        log.info({"candidates": len(candidates)}, "reconciler: not exactly one candidate")
        return

    target_invoice_id = candidates[0]
    params = {
        "payment_id": p["qbo_payment_id"],
        "invoice_id": target_invoice_id,
        "amount_cents": p["amount_cents"],
    }
    idem = compute_idempotency_key("payment.apply", event["subject_ref"], params)
    dup = find_duplicate_in_window(db, idem)
    if dup is not None:
        log.info({"idem": idem, "prior": dup.id}, "reconciler: duplicate within 24h, skipping")
        return

    db.add(
        Action(
            type="payment.apply",
            params=params,
            mode="proposed",
            drafted_by=f"agent:reconciler@{VERSION}",
            status="drafted",
            reversal_class=ACTION_CATALOG["payment.apply"].reversal_class,
            idempotency_key=idem,
            agent_confidence=str(result.proposal.confidence),
            agent_rationale=result.proposal.rationale,
            preview=f"Apply ${params['amount_cents'] / 100:.2f} to {target_invoice_id}",
            originating_event_id=event["id"],
            subject_ref=event["subject_ref"],
        )
    )
    db.add(
        AuditEvent(
            actor=f"agent:reconciler@{VERSION}",
            originating_event_id=event["id"],
            kind="agent.proposed",
            idempotency_key=idem,
            subject_ref=event["subject_ref"],
            note=result.proposal.rationale,
        )
    )
    db.commit()

File: apps/api/src/agents/reconciler.ts
import type { Event } from '@linkbook/types';
import { actionCatalog, RECONCILER_CONFIDENCE_THRESHOLD } from '@linkbook/types';
import { schema } from '@linkbook/db';
import type { AppCtx } from '../lib/principal-context.js';
import { getAgentspanClient, proposeWithFallback } from './runtime.js';
import { computeIdempotencyKey, findDuplicateInWindow } from '../idempotency.js';

const VERSION = 'v1.0';

// §5.3 — Reconciler: triggers on payment.received_unapplied. Below 0.85
// confidence the proposal is suppressed; the user resolves manually.

export async function runReconciler(event: Event, ctx: AppCtx): Promise<void> {
  if (event.type !== 'payment.received_unapplied') return;

  const client = getAgentspanClient(ctx.cfg);
  const result = await proposeWithFallback(
    client,
    {
      agent: 'reconciler',
      agent_version: VERSION,
      event,
      context: { candidate_invoice_ids: event.payload.candidate_invoice_ids },
    },
    ctx.log,
  );
  if (!result.ok) return;
  if (result.proposal.confidence < RECONCILER_CONFIDENCE_THRESHOLD) {
    ctx.log.info(
      {
        confidence: result.proposal.confidence,
        threshold: RECONCILER_CONFIDENCE_THRESHOLD,
      },
      'reconciler below threshold — leaving as manual event',
    );
    return;
  }

  // We only auto-propose when there is EXACTLY one candidate. Multiple
  // candidates always go to the user — applying to the wrong invoice is
  // the failure mode that breaks accountant trust (§4.1, §5.3).
  if (event.payload.candidate_invoice_ids.length !== 1) {
    ctx.log.info(
      { candidates: event.payload.candidate_invoice_ids.length },
      'reconciler: not exactly one candidate — leaving manual',
    );
    return;
  }
  const target_invoice_id = event.payload.candidate_invoice_ids[0];
  if (!target_invoice_id) return;

  const params = {
    payment_id: event.payload.qbo_payment_id,
    invoice_id: target_invoice_id,
    amount_cents: event.payload.amount_cents,
  };
  const idem = computeIdempotencyKey('payment.apply', event.subject_ref, params);
  const dup = await findDuplicateInWindow(ctx, idem);
  if (dup) {
    ctx.log.info({ idem, prior: dup.id }, 'reconciler: duplicate within 24h, skipping');
    return;
  }

  ctx.db
    .insert(schema.actions)
    .values({
      type: 'payment.apply',
      params,
      mode: 'proposed',
      drafted_by: `agent:reconciler@${VERSION}`,
      status: 'drafted',
      reversal_class: actionCatalog['payment.apply'].reversal_class,
      idempotency_key: idem,
      agent_confidence: String(result.proposal.confidence),
      agent_rationale: result.proposal.rationale,
      preview: `Apply $${(params.amount_cents / 100).toFixed(2)} to ${target_invoice_id}`,
      originating_event_id: event.event_id,
      subject_ref: event.subject_ref,
    })
    .run();

  // §2.6 audit row mirroring cash-chaser / concierge.
  ctx.db
    .insert(schema.audit_events)
    .values({
      actor: `agent:reconciler@${VERSION}`,
      originating_event_id: event.event_id,
      kind: 'agent.proposed',
      idempotency_key: idem,
      subject_ref: event.subject_ref,
      note: result.proposal.rationale,
    })
    .run();
}
